### ISSUE TITLE
FR: No Bare File Path

### ISSUE THREAD

**Issue by SPiCaRiA** [platers/obsidian-linter#675]
## Is Your Feature Request Related to a Problem? Please Describe.

Currently we have the `No Bare URLs` rule, which surrounds URLs with angle brackets to avoid ambiguity. I'm wondering if we can have the exact same behavior on file paths as well. Some of the use cases might be:

1. Referring an attachment
2. Referring another note (though Wiki links are recommended, some people still prefer `![]()` styled links)

If the target file has spaces in its file name, the surrounding brackets would be necessary to make things work.

## Describe the Solution You'd Like

**Please include an example where applicable:**

``` markdown
![](path/to/file-name)
![](path/to/file name with space)
```

expected outcome
``` markdown
![](<path/to/file-name>)
![](<path/to/file name with space>)
```

## Alternatives

I think an alternative way of thinking this problem is, can we provide a rule to surround angle brackets for strings inside the markdown link syntax:

```markdown
![](<enclose me with angle brackets>)
[](<and me>)
```

However, I'm not fully sure if this is something desired on URLs, given the philosophy of markdown semantics:

```markdown
[](<URL>)
![](<URL>)
```

So the choice of `No Bare File Path` vs. `Enclose All Links` needs more discussion.

**Comment by pjkaufman** [comment]
`No Bare URLs` is designed to format URLs in such a way that markdown renders it correctly for most parsers. However I am not aware of any issue where there would be a need to have a file path not be bare. The examples provided all show file paths that would already be in one markdown syntax or another which would negate a reason to put angle brackets around them.

Could you give an example of a scenario where having a space in a file path breaks the link to the file or image?

**Comment by SPiCaRiA** [comment]
> `No Bare URLs` is designed to format URLs in such a way that markdown renders it correctly for most parsers. However I am not aware of any issue where there would be a need to have a file path not be bare. The examples provided all show file paths that would already be in one markdown syntax or another which would negate a reason to put angle brackets around them.
> 
> Could you give an example of a scenario where having a space in a file path breaks the link to the file or image?

Sure! For the following directory structure:

```
ZOB_assets
└─ imgs
　　└─ WIM - Search Engine Overview.png
Test.md
```

and the markdown

```markdown
![](ZOB_assets/imgs/WIM - Search Engine Overview.png)
![](<ZOB_assets/imgs/WIM - Search Engine Overview.png>)
```

In fact, it is rendered as

<img width="705" alt="Screen Shot 2023-04-08 at 10 52 51" src="https://user-images.githubusercontent.com/11176415/230735936-c3bbc043-5f6c-44ef-9626-99e936d0bedf.png">

Note that the bare file path one is only rendered as plain text. And if you Ctrl/Cmd click on the first link under editor mode (idk why it is not plain text under editor mode tbh, may be some Obsidian render choices but anyways), it creates a new note `ZOB_assets/imgs/WIM.md`, which proves that at least for Obsidian, a file path with spaces in it is not recognized after the first space (even when it is already enclosed in the parentheses of link syntax).

Yes as you mentioned, this might be related to the parser, but at the end of the day, I'm sure angle brackets can disambiguiate the grammar.

## Alternative Solution

As I said in the original post, I'm not sure if angle brackets are the most markdown-esque way of doing this. An alternative way could be `Force Escaping Spaces in File Paths`:

**Before**
```markdown
![](ZOB_assets/imgs/WIM - Search Engine Overview.png)
```

**After**
```markdown
![](ZOB_assets/imgs/WIM%20-%20Search%20Engine%20Overview.png)
```

Which also renders.

**Comment by pjkaufman** [comment]
> > `No Bare URLs` is designed to format URLs in such a way that markdown renders it correctly for most parsers. However I am not aware of any issue where there would be a need to have a file path not be bare. The examples provided all show file paths that would already be in one markdown syntax or another which would negate a reason to put angle brackets around them.
> > Could you give an example of a scenario where having a space in a file path breaks the link to the file or image?
> 
> Sure! For the following directory structure:
> 
> ```
> ZOB_assets
> └─ imgs
> 　　└─ WIM - Search Engine Overview.png
> Test.md
> ```
> 
> and the markdown
> 
> ```gfm
> ![](ZOB_assets/imgs/WIM - Search Engine Overview.png)
> ![](<ZOB_assets/imgs/WIM - Search Engine Overview.png>)
> ```
> 
> In fact, it is rendered as
> 
> <img alt="Screen Shot 2023-04-08 at 10 52 51" width="705" src="https://user-images.githubusercontent.com/11176415/230735936-c3bbc043-5f6c-44ef-9626-99e936d0bedf.png">
> 
> Note that the bare file path one is only rendered as plain text. And if you Ctrl/Cmd click on the first link under editor mode (idk why it is not plain text under editor mode tbh, may be some Obsidian render choices but anyways), it creates a new note `ZOB_assets/imgs/WIM.md`, which proves that at least for Obsidian, a file path with spaces in it is not recognized after the first space (even when it is already enclosed in the parentheses of link syntax).
> 
> Yes as you mentioned, this might be related to the parser, but at the end of the day, I'm sure angle brackets can disambiguiate the grammar.
> 
> ## Alternative Solution
> As I said in the original post, I'm not sure if angle brackets are the most markdown-esque way of doing this. An alternative way could be `Force Escaping Spaces in File Paths`:
> 
> **Before**
> 
> ```gfm
> ![](ZOB_assets/imgs/WIM - Search Engine Overview.png)
> ```
> 
> **After**
> 
> ```gfm
> ![](ZOB_assets/imgs/WIM%20-%20Search%20Engine%20Overview.png)
> ```
> 
> Which also renders.

The example provided seems to me like it needs URI encoding and that is why it is showing up strange for you. Have you confirmed that putting angle brackets around it makes it display as desired (i.e. like a regular link)?

**Comment by SPiCaRiA** [comment]
> > > `No Bare URLs` is designed to format URLs in such a way that markdown renders it correctly for most parsers. However I am not aware of any issue where there would be a need to have a file path not be bare. The examples provided all show file paths that would already be in one markdown syntax or another which would negate a reason to put angle brackets around them.
> > > Could you give an example of a scenario where having a space in a file path breaks the link to the file or image?
> > 
> > 
> > Sure! For the following directory structure:
> > ```
> > ZOB_assets
> > └─ imgs
> > 　　└─ WIM - Search Engine Overview.png
> > Test.md
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and the markdown
> > ```
> > ![](ZOB_assets/imgs/WIM - Search Engine Overview.png)
> > ![](<ZOB_assets/imgs/WIM - Search Engine Overview.png>)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In fact, it is rendered as
> > <img alt="Screen Shot 2023-04-08 at 10 52 51" width="705" src="https://user-images.githubusercontent.com/11176415/230735936-c3bbc043-5f6c-44ef-9626-99e936d0bedf.png">
> > Note that the bare file path one is only rendered as plain text. And if you Ctrl/Cmd click on the first link under editor mode (idk why it is not plain text under editor mode tbh, may be some Obsidian render choices but anyways), it creates a new note `ZOB_assets/imgs/WIM.md`, which proves that at least for Obsidian, a file path with spaces in it is not recognized after the first space (even when it is already enclosed in the parentheses of link syntax).
> > Yes as you mentioned, this might be related to the parser, but at the end of the day, I'm sure angle brackets can disambiguiate the grammar.
> > ## Alternative Solution
> > As I said in the original post, I'm not sure if angle brackets are the most markdown-esque way of doing this. An alternative way could be `Force Escaping Spaces in File Paths`:
> > **Before**
> > ```
> > ![](ZOB_assets/imgs/WIM - Search Engine Overview.png)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **After**
> > ```
> > ![](ZOB_assets/imgs/WIM%20-%20Search%20Engine%20Overview.png)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which also renders.
> 
> The example provided seems to me like it needs URI encoding and that is why it is showing up strange for you. Have you confirmed that putting angle brackets around it makes it display as desired (i.e. like a regular link)?

Yes. Actually the second code block in my reply is an example of rendering w/ and w/o angle brackets, and the image below it is how it looks like in reader mode. Sorry that it was not trivial :x

**Comment by SPiCaRiA** [comment]
The conclusion is that Obsidian uses a grammar rule like this:

```
md-link -> [ title ]( link )
         | ! [ title ]( link )

link -> non-spatial-bare-link
      | < raw-link >
```

Where `non-spatial-bare-link` is matched by the regular expression `[^\s]*`, and `raw-link` is matched by `[^>]*`.

Then the actions are

```
// Assume link is file path.
if link != "" and link != "<>", then
    if link target exist, then
        present link or image
    else
        if link is image, then
            present md-link in plain text
        else
            present a link to create file at that location
        end
    end
end
```

TL;DR: using file paths with spaces (and not enclosed by angle brackets) will be truncated, because the regular expression matches anything **except** spaces. But when using angle brackets, the parser will keep matching until the next character is the closing angle bracket.

And as I mentioned, an alternative way is to provide a rule that matches the non-terminal `link` and replace any space with `%20` in it. So the key discussion is, when we encounter an "invalid" link (i.e., link with spaces), should we escape it for our users or wrap it up with angle brackets. Or providing an option for the user to decide which style they go with.

**Comment by vbrozik** [comment]
The reason why Markdown does not allow spaces in links is that it expects a link title after the space like `[](path/to/file "title")` See https://spec.commonmark.org/0.31.2/#example-505

...but in reality I think practically no-one uses the titles so it would certainly make sense to add an **optional** rule which would convert `[](path/to/file name with space)` to `[](<path/to/file name with space>)`. I would certainly use it. What do you think?
